### PR TITLE
Include ForeignItem when visiting types for WF check

### LINKED ISSUE #95665: `Unexpected node ForeignItem` with extern generic
### Code

```Rust
pub trait Trait: {}

pub struct Struct<T: Trait> {
    member: T,
}

// uncomment and bug goes away
//impl Trait for u8 {}

extern "C" {
    static VAR: Struct<u8>;
}

fn main() {}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.61.0-nightly (76d770ac2 2022-04-02)
binary: rustc
commit-hash: 76d770ac21d9521db6a92a48c7b3d5b2cc535941
commit-date: 2022-04-02
host: aarch64-apple-darwin
release: 1.61.0-nightly
LLVM version: 14.0.0
```

Also reproduces on 1.58.0 and 1.59.0.

### Error output

```
error: internal compiler error: compiler/rustc_typeck/src/hir_wf_check.rs:144:25: Unexpected node ForeignItem(ForeignItem { ident: VAR#0, kind: Static(Ty { hir_id: HirId { owner: DefId(0:8 ~ boom[15c8]::{extern#0}::VAR), local_id: 1 }, kind: Path(Resolved(None, Path { span: boom.rs:12:17: 12:27 (#0), res: Def(Struct, DefId(0:4 ~ boom[15c8]::Struct)), segments: [PathSegment { ident: Struct#0, hir_id: Some(HirId { owner: DefId(0:8 ~ boom[15c8]::{extern#0}::VAR), local_id: 4 }), res: Some(Err), args: Some(GenericArgs { args: [Type(Ty { hir_id: HirId { owner: DefId(0:8 ~ boom[15c8]::{extern#0}::VAR), local_id: 2 }, kind: Path(Resolved(None, Path { span: boom.rs:12:24: 12:26 (#0), res: PrimTy(Uint(U8)), segments: [PathSegment { ident: u8#0, hir_id: Some(HirId { owner: DefId(0:8 ~ boom[15c8]::{extern#0}::VAR), local_id: 3 }), res: Some(Err), args: None, infer_args: false }] })), span: boom.rs:12:24: 12:26 (#0) })], bindings: [], parenthesized: false, span_ext: boom.rs:12:23: 12:27 (#0) }), infer_args: false }] })), span: boom.rs:12:17: 12:27 (#0) }, Not), def_id: DefId(0:8 ~ boom[15c8]::{extern#0}::VAR), span: boom.rs:12:5: 12:28 (#0), vis: Spanned { node: Inherited, span: boom.rs:12:5: 12:5 (#0) } }) 

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1279:9

query stack during panic:
#0 [diagnostic_hir_wf_check] performing HIR wf-checking for predicate Binder(TraitPredicate(<u8 as Trait>, polarity:Positive), []) at item Ty(DefId(0:8 ~ boom[15c8]::{extern#0}::VAR))
#1 [check_item_well_formed] checking that `` is well-formed
#2 [analysis] running analysis passes on this crate
end of query stack

```

<details><summary><strong>Backtrace</strong></summary>

<p>

```
thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1279:9
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner>::bug
   3: <rustc_errors::Handler>::bug
   4: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, ()>
   5: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>
   6: rustc_middle::util::bug::bug_fmt
   7: rustc_typeck::hir_wf_check::diagnostic_hir_wf_check
   8: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::diagnostic_hir_wf_check, rustc_query_impl::plumbing::QueryCtxt>
   9: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::diagnostic_hir_wf_check
  10: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtExt>::report_selection_error
  11: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtPrivExt>::report_fulfillment_error
  12: <rustc_infer::infer::InferCtxt as rustc_trait_selection::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
  13: <rustc_typeck::check::fn_ctxt::FnCtxt>::select_all_obligations_or_error
  14: <rustc_infer::infer::InferCtxtBuilder>::enter::<(), <rustc_typeck::check::inherited::InheritedBuilder>::enter<<rustc_typeck::check::wfcheck::CheckWfFcxBuilder>::with_fcx<rustc_typeck::check::wfcheck::check_item_type::{closure#0}>::{closure#0}, ()>::{closure#0}>
  15: rustc_typeck::check::wfcheck::check_item_type
  16: rustc_typeck::check::wfcheck::check_item_well_formed
  17: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, ()>>
  18: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::check_item_well_formed, rustc_query_impl::plumbing::QueryCtxt>
  19: <core::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures::sync::par_for_each_in<&alloc::vec::Vec<rustc_hir::hir::MaybeOwner<&rustc_hir::hir::OwnerInfo>>, <rustc_middle::hir::map::Map>::par_visit_all_item_likes<rustc_typeck::check::wfcheck::CheckTypeWellFormedVisitor>::{closure#0}>::{closure#0}::{closure#0}> as core::ops::function::FnOnce<()>>::call_once
  20: rustc_data_structures::sync::par_for_each_in::<&alloc::vec::Vec<rustc_hir::hir::MaybeOwner<&rustc_hir::hir::OwnerInfo>>, <rustc_middle::hir::map::Map>::par_visit_all_item_likes<rustc_typeck::check::wfcheck::CheckTypeWellFormedVisitor>::{closure#0}>
  21: <rustc_session::session::Session>::track_errors::<rustc_typeck::check_crate::{closure#5}, ()>
  22: rustc_typeck::check_crate
  23: rustc_interface::passes::analysis
  24: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorGuaranteed>>>
  25: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
  26: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
  27: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorGuaranteed>>
  28: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  29: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorGuaranteed>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorGuaranteed>>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.61.0-nightly (76d770ac2 2022-04-02) running on aarch64-apple-darwin

query stack during panic:
#0 [diagnostic_hir_wf_check] performing HIR wf-checking for predicate Binder(TraitPredicate(<u8 as Trait>, polarity:Positive), []) at item Ty(DefId(0:8 ~ boom[15c8]::{extern#0}::VAR))
#1 [check_item_well_formed] checking that `` is well-formed
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to previous error
```

</p>
</details>

### PR DESCRIPTION
**Pull request by PrestonFrom**
Addresses Issue 95665 by including `hir::Node::ForeignItem` as a valid
type to visit in `diagnostic_hir_wf_check`.

Fixes #95665

### CODE REVIEW

**Comment by PrestonFrom** [comment]
This fixes the issue originally reported, but I'm wondering if the other variants in `hir::Node` should be valid to visit. The comment on line 120 *seemed* like it would justify visiting any `Ty` that could be extracted, but I definitely don't have a complete understanding of what is happening. Any guidance would be appreciated!

File: compiler/rustc_typeck/src/hir_wf_check.rs [hunk 1]
```diff
@@ -141,6 +141,11 @@ fn diagnostic_hir_wf_check<'tcx>(
                 ref item => bug!("Unexpected item {:?}", item),
             },
             hir::Node::Field(field) => Some(field.ty),
+            hir::Node::ForeignItem(item) => match item.kind {
+                ForeignItemKind::Fn(_, _, _) => None,
+                ForeignItemKind::Static(ty, _) => Some(ty),
+                ForeignItemKind::Type => None,
```
Comment (petrochenkov) [inline compiler/rustc_typeck/src/hir_wf_check.rs:147]:
```suggestion
                ref item => bug!("Unexpected foreign item {:?}", item),
```
Both `ForeignItemKind::Fn` and `ForeignItemKind::Type` should be unreachable here.

**Comment by cjgillot** [comment]
HIR WF check is just about getting a more precise diagnostic.  Can't we just let the `match` be `None` instead of panicking?  This would give us a sub-optimal diagnostic, without any chance of ICE.

**Comment by PrestonFrom** [comment]
@cjgillot Apologies if this should be obvious, but I want to confirm that you mean the match on the `ForeignItem` kind and not the "parent" match on the `WellFormedLoc`.

**Comment by cjgillot** [comment]
@PrestonFrom in the cases the inner `match` returns `None`, the outer one will return the same result, won't it?  In doubt: I mean replacing the `bug!` into `None`, not changing the other cases that already return `Some`.

**Comment by PrestonFrom** [comment]
@cjgillot The outer match currently uses `bug!` for the non-matching case, which I think is why it was suggested to use `bug!` here as well. So, I wasn't sure if you wanted to update both to return `None` or just the new addition.

File: compiler/rustc_typeck/src/hir_wf_check.rs (hunk 1 above)
Comment (PrestonFrom) [reply compiler/rustc_typeck/src/hir_wf_check.rs]: @petrochenkov Sorry, I missed your suggestion and ended up constraining the match instead. Is it better to match ForeignType and then match on its kind?

Comment (petrochenkov) [reply compiler/rustc_typeck/src/hir_wf_check.rs]: The current version is ok too.

**Comment by petrochenkov** [comment]
r=me after squashing commits.

**Comment by PrestonFrom** [comment]
r? @petrochenkov 
Commits squashed and pushed

**Comment by PrestonFrom** [comment]
r=@petrochenkov

**Comment by petrochenkov** [comment]
Thanks!
@bors r+ rollup

**Pull request merged**